### ISSUE TITLE
Support for Nodejs

### ISSUE THREAD

**Issue by vpavlin** [codex-storage/codex-js#8]
I tried to ise codex-js in nodejs (while playing around with @ai16z/eliza) and attempt to upload data failed with an error about `xmlhttprequest is not defined` in nodejs

**Comment by 2-towns** [comment]
Yes this is expected because xmlhttprequest is only available in the browser. The solution would be to use fetch but without the progress callback support. Do you have a repo to share ?

**Comment by 2-towns** [comment]
Duplicate #3

**Comment by vpavlin** [comment]
Ah, sorry missed the existing issue 🤦‍♂️

When I get a cance, I'll create a reproducer. I think maybe switching between fetch and xmlhttprequest based on whether progress callback is passed + whether in browser would make sense? (Although it would add some maintenance cost, but I feel like we need codex-js to work with Node)

**Comment by 2-towns** [comment]
> I think maybe switching between fetch and xmlhttprequest based on whether progress callback is passed + whether in browser would make sense?

Yes that's an option. But I prefer to make an explicit option rather than auto-detection, it give more flexibility. The default  would be xmlhttprequest if it is in the browser + callback is passed.

**Comment by vpavlin** [comment]
@2-towns did you get a chance to get to this?:)

**Comment by 2-towns** [comment]
> @2-towns did you get a chance to get to this?:) 

I didn’t have a clear use case until now, so it wasn’t a priority. But now that there is one, I’ll work on it soon.

**Comment by 2-towns** [comment]
@vpavlin can you try this branch please https://github.com/codex-storage/codex-js/tree/feat/node-js-support ? 

To make compatible for browser and Node JS, I introduced [browser strategy](https://github.com/codex-storage/codex-js/blob/7b8fcc449b23909117dbab04d66789b5a42aba5b/README.md?plain=1#L281) and [node strategy](https://github.com/codex-storage/codex-js/blob/7b8fcc449b23909117dbab04d66789b5a42aba5b/README.md?plain=1#L311). 

You can check the examples folder as well and of course, I am available if you need any support.

**Comment by vpavlin** [comment]
```
describe("Test Codex", () => {
  it("should be able to create a Codex instance", async () => {
    const codex = new CodexStorageProvider({ codexURI: process.env.CODEX_URI});
    const cid = await codex.upload("output.md", "Hello World")
    console.log(cid)
    const content = await codex.download(cid)

    expect(content).to.eq("Hello World")
    console.log(content)
 
  });
});
```

I was able to use local build of codex-js from the mentioned branch in the `filevere-agents` and run this test successfully via nodejs:-)

**Comment by 2-towns** [comment]
> Yes, I was able to upload an object using the node strategy. Seems @types/node dependency is missing (at least compile errors resolved after installing it:))

Yes this is something that you should have in your `devDependencies` , I should not include it into the sdk `dependencies`.

**Comment by 2-towns** [comment]
> ~Is this just for uploading at the moment?~
> 
> Nevermind, my fingers are faster than my brain today:-)

Just to make sure we are on the same page, the problem was only with the upload, because it is using `XMLHttpRequest` to have the progress information. 

So everything else *should* work.

**Comment by 2-towns** [comment]
@vpavlin if it is okay for you, I can make a new release.

**Comment by vpavlin** [comment]
Ok, that's fine, I am not sure why it is missing from `fileverse-agents`, but that is a problem for another day:-)

> So everything else _should_ work.

Ah, makes sense then:-) All good, the upload/download flow seems to work fine in nodejs now. I have not tested the browser flow - is this a breaking change or will you assume browser is no explicit strategy is specified (sorry did not look tha tdeep in your changes)

**Comment by 2-towns** [comment]
> Ah, makes sense then:-) All good, the upload/download flow seems to work fine in nodejs now. I have not tested the browser flow - is this a breaking change or will you assume browser is no explicit strategy is specified (sorry did not look tha tdeep in your changes)

It is a breaking change. I added a note in the [README](https://github.com/codex-storage/codex-js/blob/7b8fcc449b23909117dbab04d66789b5a42aba5b/README.md?plain=1#L11).

**Comment by vpavlin** [comment]
Ok, sounds good, just wanted to make sure as I should probably also update Qaku:-) THanks a lot! LGTM!
